microcosm_flask/conventions/upload.py

```diff
@@ -0,0 +1,136 @@
+"""
+Conventions for file upload.
+
+"""
+from contextlib import contextmanager
+from os.path import join
+from shutil import rmtree
+from tempfile import mkdtemp
+
+from flask import request
+from marshmallow import Schema
+from microcosm_flask.conventions.base import Convention
+from microcosm_flask.conventions.encoding import (
+    dump_response_data,
+    load_query_string_data,
+    merge_data,
+)
+from microcosm_flask.conventions.registry import qs, response
+from microcosm_flask.operations import Operation
+from werkzeug.exceptions import BadRequest
+from werkzeug.utils import secure_filename
+
+
+try:
+    # Python 2
+    from contextlib import nested
+except ImportError:
+    # Python3
+    from contextlib import ExitStack
+
+    @contextmanager
+    def nested(*contexts):
+        """
+        Reimplementation of nested in python 3.
+        """
+        with ExitStack() as stack:
+            results = [
+                stack.enter_context(context)
+                for context in contexts
+            ]
+            yield results
+
+
+@contextmanager
+def temporary_upload(name, fileobj):
+    """
+    Upload a file to a temporary location.
+
+    Flask will not load sufficiently large (>500M) files into memory, so it
+    makes sense to always load files into a temporary directory.
+
+    """
+    tempdir = mkdtemp()
+    filename = secure_filename(fileobj.filename)
+    filepath = join(tempdir, filename)
+    fileobj.save(filepath)
+    yield name, filepath, fileobj.filename
+    rmtree(tempdir)
+
+
+class UploadConvention(Convention):
+
+    def __init__(self, graph, exclude_func=None):
+        self.graph = graph
+        self.exclude_func = exclude_func or (lambda name, fileobj: False)
+
+    def create_upload_func(self, ns, definition, path, operation):
+        request_schema = definition.request_schema or Schema()
+        response_schema = definition.response_schema or Schema()
+
+        @self.graph.route(path, operation, ns)
+        def upload(**path_data):
+            request_data = load_query_string_data(request_schema)
+
+            if not request.files:
+                raise BadRequest("No files were uploaded")
+
+            uploads = [
+                temporary_upload(name, fileobj)
+                for name, fileobj
+                in request.files.items()
+                if not self.exclude_func(name, fileobj)
+            ]
+            with nested(*uploads) as files:
+                response_data = definition.func(files, **merge_data(path_data, request_data))
```

Comment:
We pass a list of tuples containing the multipart form's field name, the uploaded file path, and the uploaded file's original name.

We also pass the path and query string data.